From 1f2f5464e88483fbe449f762a4c0bf1972889c0f Mon Sep 17 00:00:00 2001
From: Erik Schumacher <erik.schumacher@iris-sensing.com>
Date: Wed, 10 Apr 2024 11:12:35 +0200
Subject: [PATCH] media/imx: Fix parallel camera and media drivers

- Fix missing first 3 frames due to vsync counter
- Allow format code propagation to sub devices instead of hardcoded MEDIA_BUS_FMTs

Upstream-Status: Pending

Signed-off-by: Erik Schumacher <erik.schumacher@iris-sensing.com>
---
 drivers/staging/media/imx/imx8-isi-cap.c      | 2 +-
 drivers/staging/media/imx/imx8-isi-hw.c       | 2 +-
 drivers/staging/media/imx/imx8-parallel-csi.c | 4 ++--
 3 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/staging/media/imx/imx8-isi-cap.c b/drivers/staging/media/imx/imx8-isi-cap.c
index d823d7a9715c..76cf7d7f5745 100644
--- a/drivers/staging/media/imx/imx8-isi-cap.c
+++ b/drivers/staging/media/imx/imx8-isi-cap.c
@@ -958,7 +958,7 @@ static int mxc_isi_source_fmt_init(struct mxc_isi_cap_dev *isi_cap)
 
 	src_fmt.pad = source_pad->index;
 	src_fmt.which = V4L2_SUBDEV_FORMAT_ACTIVE;
-	src_fmt.format.code = MEDIA_BUS_FMT_UYVY8_1X16;
+	src_fmt.format.code = dst_f->fmt->mbus_code;
 	src_fmt.format.width = dst_f->width;
 	src_fmt.format.height = dst_f->height;
 	ret = v4l2_subdev_call(src_sd, pad, set_fmt, NULL, &src_fmt);
diff --git a/drivers/staging/media/imx/imx8-isi-hw.c b/drivers/staging/media/imx/imx8-isi-hw.c
index f3b2b98798be..29aab0911702 100644
--- a/drivers/staging/media/imx/imx8-isi-hw.c
+++ b/drivers/staging/media/imx/imx8-isi-hw.c
@@ -673,7 +673,7 @@ void mxc_isi_channel_enable_loc(struct mxc_isi_dev *mxc_isi, bool m2m_enabled)
 	u32 val;
 
 	val = readl(mxc_isi->regs + CHNL_CTRL);
-	val |= 0xff << CHNL_CTRL_BLANK_PXL_OFFSET;
+	val |= 0x3f << CHNL_CTRL_BLANK_PXL_OFFSET;
 
 	if (m2m_enabled) {
 		val &= ~(CHNL_CTRL_SRC_TYPE_MASK | CHNL_CTRL_SRC_INPUT_MASK);
diff --git a/drivers/staging/media/imx/imx8-parallel-csi.c b/drivers/staging/media/imx/imx8-parallel-csi.c
index 3cbe39d509d2..a25ef807b025 100644
--- a/drivers/staging/media/imx/imx8-parallel-csi.c
+++ b/drivers/staging/media/imx/imx8-parallel-csi.c
@@ -393,11 +393,12 @@ static void mxc_pcsi_csr_config(struct mxc_parallel_csi_dev *pcsidev)
 	/* Config CTRL REG */
 	val = readl(pcsidev->csr_regs + pdata->interface_ctrl_reg);
 
+	val &= ~(CSI_CTRL_REG_MASK_VSYNC_COUNTER(3));
 	val |= (CSI_CTRL_REG_DATA_TYPE_IN(pdata->def_csi_in_data_type) |
 		pdata->def_hsync_pol << CSI_CTRL_REG_HSYNC_POL_OFFSET |
 		pdata->def_vsync_pol << CSI_CTRL_REG_VSYNC_POL_OFFSET |
 		pdata->def_pixel_clk_pol << CSI_CTRL_REG_PIXEL_CLK_POL_OFFSET |
-		CSI_CTRL_REG_MASK_VSYNC_COUNTER(3) |
+		CSI_CTRL_REG_MASK_VSYNC_COUNTER(0) |
 		CSI_CTRL_REG_HSYNC_PULSE(2));
 
 	if (pcsidev->uv_swap)
@@ -657,7 +658,6 @@ static int mxc_pcsi_set_fmt(struct v4l2_subdev *sd,
 	}
 
 	fmt->pad = source_pad->index;
-	fmt->format.code = MEDIA_BUS_FMT_UYVY8_2X8;
 	ret = v4l2_subdev_call(sen_sd, pad, set_fmt, NULL, fmt);
 	if (ret < 0 && ret != -ENOIOCTLCMD)
 		return ret;
-- 
2.46.0

